index.ts

```diff
@@ -11,58 +11,69 @@ type Options<T> = Partial<{
   syncData: boolean;
 }>;
 
-function useLocalStorage<T>(
-  key: string,
-  defaultValue: T,
-  options?: Options<T>
-): [T, Setter<T>];
 function useLocalStorage<T>(
   key: string,
   defaultValue?: T,
   options?: Options<T>
-) {
+): [T, Setter<T>] {
   const opts = useMemo(() => {
     return {
       serializer: JSON.stringify,
       parser: JSON.parse,
       logger: console.log,
       syncData: true,
-      ...options,
+      ...options
     };
   }, [options]);
 
   const { serializer, parser, logger, syncData } = opts;
+  // The parser and logger are supposed to change very infrequently since
+  // they are memoized above; however the key and default value are not
+  // memoized in any way so we take them as params and expect the caller
+  // of this function to properly depend on them
+  const getValue = useCallback(
+    (key, defaultValue) => {
+      if (typeof window === "undefined") return defaultValue;
+      try {
+        const item = window.localStorage.getItem(key);
+        const res = item ? parser(item) : defaultValue;
 
-  const [storedValue, setValue] = useState(() => {
-    if (typeof window === "undefined") return defaultValue;
+        return res;
+      } catch (e) {
+        logger(e);
+        return defaultValue;
+      }
+    },
+    [parser, logger]
+  );
 
-    try {
-      const item = window.localStorage.getItem(key);
-      const res: T = item ? parser(item) : defaultValue;
-      return res;
-    } catch (e) {
-      logger(e);
-      return defaultValue;
-    }
-  });
+  const [storedValue, setValue] = useState(() => getValue(key, defaultValue));
+
+  // When any of the dependencies is updated, check
+  useEffect(() => {
+    setValue(getValue(key, defaultValue));
+  }, [getValue, key, defaultValue]);
 
   useEffect(() => {
     if (typeof window === "undefined") return;
 
     const updateLocalStorage = () => {
-      if (storedValue !== undefined) {
-        window.localStorage.setItem(key, serializer(storedValue));
-      } else {
+      if (storedValue === undefined) {
```

Comment:
Removed double negative; if the new value is undefined, then it should be set to the default value.